Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -83,74 +87,166 @@ public async Task<BoolResult> ShutdownAsync(Context context)
         {
-            try
-            {
-                using (var stream = new FileStream(destinationPath.Path, FileMode.Create, FileAccess.Write, FileShare.None, ContentStore.Grpc.Utils.DefaultBufferSize, FileOptions.Asynchronous | FileOptions.SequentialScan))
-                {
-                    return await CopyToAsync(context, contentHash, stream, cancellationToken, fileSize, enableCompression);
-                }
-            }
-            catch (Exception ex) when (ex is UnauthorizedAccessException || ex is InvalidOperationException)
-            {
-                return new CopyFileResult(CopyFileResult.ResultCode.DestinationPathError, ex, $"Error copying to file {destinationPath.Path}");
-            }
+            Func<Stream> streamFactory = () => new FileStream(destinationPath.Path, FileMode.Create, FileAccess.Write, FileShare.None, ContentStore.Grpc.CopyConstants.DefaultBufferSize, FileOptions.Asynchronous | FileOptions.SequentialScan);
+            return CopyToAsync(context, hash, streamFactory, ct);        
         }
-
+        
         /// <summary>
-        /// Copies file to stream
+        /// Copies content from the server to the given stream.
         /// </summary>
-        public async Task<CopyFileResult> CopyToAsync(Context context, ContentHash contentHash, Stream stream, CancellationToken cancellationToken, long fileSize = -1, bool enableCompression = false)
+        public Task<CopyFileResult> CopyToAsync(Context context, ContentHash hash, Stream stream, CancellationToken ct = default(CancellationToken))
         {
-            // TODO: Pipe through flag for compression type
-            CopyCompression compression = enableCompression ? CopyCompression.Gzip : CopyCompression.None;
-            long bytesReceived = 0L;
+            return CopyToAsync(context, hash, () => stream, ct);
+        }
 
+        /// <summary>
+        /// Copies content from the server to the stream returned by the factory.
+        /// </summary>
+        public async Task<CopyFileResult> CopyToAsync(Context context, ContentHash hash, Func<Stream> streamFactory, CancellationToken ct = default(CancellationToken))
+        {
             try
             {
-                AsyncServerStreamingCall<CopyFileResponse> response = _client.CopyFile(new CopyFileRequest
+                CopyFileRequest request = new CopyFileRequest()
                 {
                     TraceId = context.Id.ToString(),
-                    HashType = (int)contentHash.HashType,
-                    ContentHash = contentHash.ToByteString(),
-                    // TODO: If `Drive` is expected to be the drive of the file on the source machine, then this should have nothing to do with the destination's drive
-                    Drive = "B",
+                    HashType = (int) hash.HashType,
+                    ContentHash = hash.ToByteString(),
+                    Drive = "X",

```

Comment:
> "X" [](start = 27, length = 4)

Not sure what this one means... Just checking this isn't `X:\` on windows... :)

Reply:
Drive isn't used, but I didn't want to break calling contract by removing it. Will revisit in next iteration.

Reply:
Went ahead and dealt with removing drive entirely in next iteration.